Blocks: Add DC Spike Removal and IQ Swap Native Blocks

ghostop14 opened the pull request:
This block incorporates the OOT modules in correctiq that provide
3 different techniques to remove the DC spike inherent in IQ sampling.
The first technique mirrors SDR GUI receivers with an IIR filter
approach, the second provides a time-limited IIR approach where after
a user-configurable number of seconds locks into a basic offset
correction to eliminate the effect of the filter on the signal while
maintaining the correction, and the last is manual I and Q
configurable offsets.  The Swap IQ block is a drop-in block to help
correct for inverted spectrums and just swaps I<->Q.

ghostop14 commented:
@mbr0wn let me know if the block names in the UI sound good.  I think the original "correctiq" names didn't convey to a user what the blocks did on their own so I gave them more appropriate names, but open to any suggestions.

marcusmueller commented on gr-blocks/include/gnuradio/blocks/correctiq.h:
```diff
@@ -0,0 +1,44 @@
+namespace blocks {
+
+/*!
+ * \brief <+description of block+>
```
:) I love this commit so far. Could you copy & paste the content from the YAML's doc field here?

marcusmueller commented on gr-blocks/include/gnuradio/blocks/correctiq_auto.h:
```diff
@@ -0,0 +1,51 @@
+namespace blocks {
+
+/*!
+ * \brief <+description of block+>
```
a short word on how this block differs from the general one, and from the other implementations, would be most helpful here.

marcusmueller commented on gr-blocks/lib/correctiq_auto_impl.cc:
```diff
@@ -0,0 +1,123 @@
+
+    synchronized = false;
+
+    std::cout << "CorrectIQ Auto Synchronizing..." << std::endl;
```
ah! Could you use replace that with 

```
GR_LOG_INFO(d_logger,  "Auto Synchronizing..."); 
```

That way, you get configurable logging. 

(the logger should inherently also print the block's name)

marcusmueller commented on gr-blocks/lib/correctiq_auto_impl.cc:
```diff
@@ -0,0 +1,123 @@
+
+void correctiq_auto_impl::set_freq(float new_value)
+{
+    std::cout << "[CorrectIQ_Auto] Auto Synchronizing..." << std::endl;
```
same :) let's unify our logging.

marcusmueller commented on gr-blocks/lib/correctiq_auto_impl.cc:
```diff
@@ -0,0 +1,123 @@
+
+void correctiq_auto_impl::set_gain(float new_value)
+{
+    std::cout << "[CorrectIQ_Auto] Auto Synchronizing..." << std::endl;
```
should these all be the same message?

marcusmueller commented on gr-blocks/lib/correctiq_auto_impl.cc:
```diff
@@ -0,0 +1,123 @@
+        std::cout << "CorrectIQ Auto offset now synchronized." << std::endl;
+        std::cout << "Applying these offsets... Real:" << std::fixed << std::setw(11)
+                  << std::setprecision(6) << -avg_real << ", imag:" << -avg_img
+                  << std::endl;
```
when logging formatted strings, you can either use `std::stringstream` to generate a string, or, very popular, `boost::format()`, which might be a bit easier to use. 

I really like that this generates a lot of debug info – but that really illustrates how important it is to keep cout clean in case someone's looking for something else in there.

marcusmueller commented on gr-blocks/lib/correctiq_swapiq_impl.cc:
```diff
@@ -0,0 +1,92 @@
+    d_datatype = datatype;
+
+    if (d_datatype != SWAPTYPE_FLOATCOMPLEX) {
+        gr::block::set_output_multiple(2); // Make sure we work with pairs
```
YES :) this is elegant.

marcusmueller commented:
Also, unit tests would be grand. Do you think you have the bandwidth to add such?

(idea for the manual IQ correction, by the way: add message handlers to set things that just call your setters; that would make the block more versatile with an offset detector downstream)

ghostop14 commented:
Hi @marcusmueller, thanks for the review.  Yes I'll have some bandwidth early this coming week.  I'll update the comments and take a look at what message ports make sense.  I've never done a unit test for a block before but I'll take a look through some of the existing ones.

mbr0wn commented on gr-blocks/lib/correctiq_auto_impl.cc:
```diff
@@ -0,0 +1,123 @@
+correctiq_auto_impl::correctiq_auto_impl(float samp_rate,
+                                         float freq,
+                                         float gain,
+                                         float syncWindow)
```
:camel:

marcusmueller replied:
@mbr0wn dromedary?

dkozel replied:
Camel Case variable. `syncWindow` should be `sync_window`

ghostop14 replied:
Lol, yeah I thought I search/replaced 'em all out.  Guess I missed one.  Old training.... snake_case was legacy c code and embedded, and anything OO was to be camel.  If you look at gr-correctiq, I caught most of them! :)

ghostop14 commented:
@marcusmueller I still have to look at the unit tests tomorrow but I had a little time to go through the other requested changes.  Everything's in the new push.  Message ports were added to the manual block and yml comments integrated.  One camel variable also fixed.  I also added some messages to the autosync block based on your manual comments.  There's now an inbound message to allow other blocks to manually trigger a resync (can be used with freq/gain as messages or any other upstream reason), and output messages that indicate when a sync is in progress, then an output with the calculated offsets when the sync is complete.  Let me know if these hit the mark.

marcusmueller commented on gr-blocks/lib/correctiq_auto_impl.cc:
```diff
@@ -0,0 +1,179 @@
+    d_max_syncSamples = (long)(d_samp_rate * (double)sync_window);
+
+    message_port_register_in(pmt::mp("rsync"));
+    set_msg_handler(pmt::mp("rsync"),
```
oh, nice use of a lambda!

marcusmueller commented on gr-blocks/lib/correctiq_auto_impl.cc:
```diff
@@ -0,0 +1,179 @@
+{
+    pmt::pmt_t meta = pmt::make_dict();
+
+    meta = pmt::dict_add(meta, pmt::mp("real"), pmt::mp(avg_real));
```
Just a warning: `pmt::mp` calls `pmt::symbol` here, and that's relatively expensive (yes, that's stupid.).

One solution I've found for that is to have a static variable in the function you use the symbol, so that the symbol construction only happens once.

For now, this is definitely not a blocker. Especially since `send_sync_value` is not expected to be called extremely often.

marcusmueller commented on gr-blocks/lib/correctiq_man_impl.cc:
```diff
@@ -0,0 +1,127 @@
+
+            set_real(new_value);
+        } else {
+            GR_LOG_INFO(d_logger, "ERROR: non-float real value received.");
```
you could use `GR_LOG_ERROR` instead of `_INFO` here :) but this is not a clear-cut case. This might really be an error, or you foresee an application where this might "just happen", in which case `_INFO` would be right.

marcusmueller commented on gr-blocks/include/gnuradio/blocks/correctiq.h:
```diff
@@ -0,0 +1,44 @@
+namespace blocks {
+
+/*!
+ * \brief This block to removes that center frequency IQ DC spike with an IIR filter.
```
Excellent.

marcusmueller commented:
> Let me know if these hit the mark.

They absolutely do, @ghostop14!

I'd instantly merge this, but I think if I leave this open, you could add the tests later, and then we have one nice round merge :)

ghostop14 commented:
@marcusmueller I took a look through some of the test cases and for the rotator for example I like the test cpu versus volk.  For these blocks I think we’d just end up with the same repeated logic and an always-match result.  So I think we should be go with the blocks.

michaelld commented on gr-blocks/grc/blocks_correctiq.block.yml:
```diff
@@ -0,0 +1,20 @@
+id: blocks_correctiq
```
Need to add these 4 blocks to the file `blocks.tree.yml`.

ghostop14 commented:
@michaelld Added.  But I'm curious for future reference.  The CMakeLists.txt in the GRC directory has a file(GLOB yml_files "*.yml") which picked up the new yml files and they did show up correctly in the UI before adding them to the blocks.tree.yml.  Where does the blocks.tree.yml come into play?

ghostop14 commented:
@michaelld I pushed another tweak.  The tree part wasn't necessary in my original push because I had the category set in the yml.  But it didn't match the convention in the gr-blocks so I removed the yml category in my files and now it just relies on blocks.tree.yml to set it like the rest of the gr-blocks.  Should be standardized now.

michaelld commented:
> @michaelld Added. But I'm curious for future reference. The CMakeLists.txt in the GRC directory has a file(GLOB yml_files "*.yml") which picked up the new yml files and they did show up correctly in the UI before adding them to the blocks.tree.yml. Where does the blocks.tree.yml come into play?

LOL I honestly don't know! But, it seems like it's desirable to keep the YML in there ... hence my ask!

michaelld commented on gr-blocks/lib/correctiq_swapiq_impl.cc:
```diff
@@ -0,0 +1,92 @@
+            out_byte[i + 1] = in_byte[i];
+        }
+    } break;
+    }
```
no `default:`? Thinking it's wise to add that, and it throws an error that the data size is not usable / known.

michaelld commented on gr-blocks/lib/correctiq_swapiq_impl.h:
```diff
@@ -0,0 +1,41 @@
+
+#include <gnuradio/blocks/correctiq_swapiq.h>
+
+#define SWAPTYPE_FLOATCOMPLEX 1
```
not a strong opinion, but an `enum` might be more robust here than these macros.

ghostop14 commented:
@michaelld I got most of the comments integrated.  Changed the errors to warns and used GR_LOG_WARN instead and added the "ignoring".  Also changed the out_float to *out_float++, and added the default to the switch that just WORK_DONE's if we don't know how to handle it.  I'm a little leary of getting too fancy with the ++ -- on the other side of the pointers (too much risk on screwing up the logic).

michaelld commented on gr-blocks/lib/correctiq_auto_impl.cc:
```diff
@@ -0,0 +1,179 @@
+
+    synchronized = false;
+
+    GR_LOG_INFO(d_logger, "Auto Synchronizing...");
```
you have 3-4 places that print this note. What I think would be more useful is to augment each with why the auto sync is happening. Here it would be "At construction; Auto Synchronizing" or maybe "Auto Synchronizing @ Construction". Another could be "Got resync command; Auto Synchronizing" ... and so forth. Just makes the note more useful.

michaelld commented on gr-blocks/lib/correctiq_auto_impl.cc:
```diff
@@ -0,0 +1,179 @@
+    gr::thread::scoped_lock guard(d_mutex);
+
+    GR_LOG_INFO(d_logger, "Auto Synchronizing...");
+    synchronized = false;
```
Since this code seems to be common here and in `set_freq` and `set_gain`, I'd think making a private method to handle the common settings would make sense. That would be just 1 place to make the same change if this block is ever augmented or those settings need to be changed ... rather than 3.

michaelld commented on gr-blocks/lib/correctiq_auto_impl.cc:
```diff
@@ -0,0 +1,179 @@
+    gr_complex* out = (gr_complex*)output_items[0];
+
+    for (int i = 0; i < noutput_items; i++) {
+        if (synchronized) {
```
I think this can be tightened up to:
```
        if (! synchronized) {
            avg_real = ratio * (in[i].real() - avg_real) + avg_real;
            avg_img = ratio * (in[i].imag() - avg_img) + avg_img;
            sync_counter++;
        }
        out[i].real(in[i].real() - avg_real);
        out[i].imag(in[i].imag() - avg_img);
```
This last line might be more clearly written instead as:
```
out[i] = in[i] - gr_complex (avg_real, avg_imag);
```
though I'm not sure what this does to the compiled code in terms of complexity.

michaelld commented on gr-blocks/lib/correctiq_auto_impl.cc:
```diff
@@ -0,0 +1,179 @@
+        }
+    }
+
+    if (!synchronized && (sync_counter >= d_max_syncSamples)) {
```
Do you want this check to be above where `sync_counter++` happens? Seems like if the "max" number of sync Samples is set, we don't want to go beyond it .. or maybe that's OK?

michaelld commented on gr-blocks/lib/correctiq_auto_impl.cc:
```diff
@@ -0,0 +1,179 @@
+    const gr_complex* in = (const gr_complex*)input_items[0];
+    gr_complex* out = (gr_complex*)output_items[0];
+
+    for (int i = 0; i < noutput_items; i++) {
```
the overall computation is just "out = in - const(avg_real, avg_imag);", regardless of whether synched or not. When not in sync, the "const" is changing, and hence the loop is required. But, when in sync, seems like you could create a Volk vectors and then use `volk_32fc_x2_add_32fc` to add them together. You'd have to keep track of the Volk vectors to make sure they are long enough, which does add a touch of complexity, but I tend to believe it would be a lot faster than doing this item by item. In the constructor you could use `set_max_noutput_items` to some reasonable amount and allocate the Volk buffers to that value (in items). Just thinking of alternative optimizations to this algorithm that wouldn't be too difficult to implement.

michaelld commented on gr-blocks/lib/correctiq_man_impl.cc:
```diff
@@ -0,0 +1,127 @@
+    const gr_complex* in = (const gr_complex*)input_items[0];
+    gr_complex* out = (gr_complex*)output_items[0];
+
+    for (int i = 0; i < noutput_items; i++) {
```
Guessing this would be faster using Volk.

michaelld replied:
Do you want to Volk-ify this? I'm happy to run you through the steps ... this one is easy-peasy & would be a lot faster!

ghostop14 commented:
@michaelld, few more tweaks there.  I updated the sync messages and put a common function in.  The reason code comes in as a param so we can still get the reason for the resync out.  I also incorporated the !synchronized optimization (I like folks are thinking optimizations!)

On the sync_counter++ question, yes.  That's how the time limit happens.  It calculates samples based on time and sample rate in the constructor to know when to time out and I can track it with the counter rather than timers.

On the "out[i] = in[i] - gr_complex (avg_real, avg_imag)" question, it was a design choice.  I was trying to avoid creating a new temporary variable every iteration.  So rather than a new complex then an assignment operator call, I go straight to setting the already-created variable.  Unfortunately the complex class doesn't have a call to set both the real and complex at the same time.  Would be nice if they did.  Unless you think the compiler would somehow optimize it out?  Given the functions available in the complex class though I think it would be slower.

On the volk side, I love when there's a volk kernel but I don't have the skillz to add the new proto-kernels.  If we have some that might match that would improve the performance that would be great.  I had timed these early on, and they're already really high throughput so I'm not as worried as say with the Costas loop one (although no need to be inefficient if there's an alternative).  Otherwise maybe a volk feature request and just track to incorporate later?

michaelld commented on gr-blocks/lib/correctiq_impl.cc:
```diff
@@ -0,0 +1,63 @@
+                     gr::io_signature::make(1, 1, sizeof(gr_complex)),
+                     gr::io_signature::make(1, 1, sizeof(gr_complex)))
+{
+    avg_real = 0.0;
```
these should be moved to either the instantiation list (just above, right after the `gr::sync_block` or into the `_impl.h` class itself.

michaelld replied:
oh ... you've already set these to 0.0 in the `_impl.h` ... so these lines here aren't required at all.

michaelld commented on gr-blocks/lib/correctiq_man_impl.h:
```diff
@@ -0,0 +1,48 @@
+class correctiq_man_impl : public correctiq_man
+{
+private:
+    float avg_real = 0.0;
```
`d_`

ghostop14 commented:
Incorporated all of @michaelld's feedback.  Cleaned up vars, man and auto now make use of volk.

michaelld commented on gr-blocks/lib/correctiq_impl.cc:
```diff
@@ -0,0 +1,63 @@
+                     gr::io_signature::make(1, 1, sizeof(gr_complex)),
+                     gr::io_signature::make(1, 1, sizeof(gr_complex)))
+{
+    avg_real = 0.0;
```
You can remove these 2 settings, since they are already done in the `_impl.h` class.

michaelld commented on gr-blocks/lib/correctiq_impl.h:
```diff
@@ -0,0 +1,38 @@
+class correctiq_impl : public correctiq
+{
+private:
+    float avg_real = 0.0;
```
prefer `d_` for class variables

ghostop14 commented:
@michaelld All standardized.  Also fixed an initialization error in the auto block complaining about out-of-order initialization.  I think this is good to go too.

michaelld commented:
@ghostop14 do you have any QA for these? that's the one thing missing here. not sure how critical it is, but if you have some already it'd be good to get that in here.

ghostop14 commented:
I don't have any QA per se on these.  I use the flowgraphs for testing them out and I've used them in live flowgraphs.  Any QA would probably only look like a 1==1 style test.  Marcus had asked and I looked into it and it didn't look like anything that could be practically implemented to provide any validation.  So I think this is good as is short any additional requests.

michaelld commented:
CI passes (even though this PR doesn't add tests, at least all current tests pass). Multiple positive reviews. Merging.
